### PR TITLE
Remove deprecated warning in synthetic def

### LINKED ISSUE #11022: deprecated case class warns at its definition
## Minimized code

```Scala
@deprecated("no CaseClass", "0.1") case class CaseClass(rgb: Int)
```

## Output

```scala
-- Deprecation Warning: CaseClass.scala:1:0 ----
21 |@deprecated("no CaseClass", "0.1") case class CaseClass(rgb: Int)
   |^
   |class CaseClass is deprecated since 0.1: no CaseClass
-- Deprecation Warning: CaseClass.scala:1:46 ---
21 |@deprecated("no CaseClass", "0.1") case class CaseClass(rgb: Int)
   |                                              ^
   |                   class CaseClass is deprecated since 0.1: no CaseClass
```

## Expectation

no warning as in Scala 2.13.4.

### PR DESCRIPTION
**Pull request by changvvb**
Fixes #11022

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -905,16 +905,19 @@ object RefChecks {
...
      * for a given declaring scope:
-     * - the symbol associated with the scope is also deprecated.
+     * - the symbol associated with the scope is also deprecated or synthetic real method.
      * - if and only if `sym` is an enum case, the scope is either
      *   a module that declares `sym`, or the companion class of the
      *   module that declares `sym`.
      */
     def skipWarning(using Context) =
-      ctx.owner.ownersIterator.exists(if sym.isEnumCase then isDeprecatedOrEnum else _.isDeprecated)
+      if sym.isEnum then ctx.owner.ownersIterator.exists(isDeprecatedOrEnum)
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:919]: I believe it is still simpler to put the if condition inside the argument to `exists`

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 2]
```diff
@@ -905,16 +905,19 @@ object RefChecks {
       owner.isDeprecated
       || isEnumOwner(owner)
 
+    def isDeprecatedOrSyntheticMethod(owner: Symbol) = owner.isDeprecated || (owner.isRealMethod && owner.is(Synthetic))
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:908]: a call to this method should replace the first condition of `isDeprecatedOrEnum`

**Review by bishabosha — CHANGES_REQUESTED** [review]
Thank you for opening this PR, I will link to this comment on why we need to be careful with ignoring synthetic defs which I see you already protect against some cases: https://github.com/lampepfl/dotty/pull/10996#issuecomment-754680777

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 3]
```diff
@@ -916,8 +916,7 @@ object RefChecks {
      *   module that declares `sym`.
      */
     def skipWarning(using Context) =
-      if sym.isEnum then ctx.owner.ownersIterator.exists(isDeprecatedOrEnum)
-      else ctx.owner.ownersIterator.exists(isDeprecatedOrSyntheticMethod)
+      ctx.owner.ownersIterator.exists(if sym.isEnumCase then isDeprecatedOrEnum else isDeprecatedOrSyntheticMethod)
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:919]: `isDeprecatedOrEnum` still needs to call `isDeprecatedOrSyntheticMethod`, the condition here was just to lift the check out of the loop

**Comment by changvvb** [comment]
> Thank you for opening this PR, I will link to this comment on why we need to be careful with ignoring synthetic defs which I see you already protect against some cases: https://github.com/lampepfl/dotty/pull/10996#issuecomment-754680777

@bishabosha I want to know some cases that we should warn in synthetic scope

**Comment by smarter** [comment]
> I want to know some cases that we should warn in synthetic scope

I mentioned some cases in https://github.com/lampepfl/dotty/pull/10996#issuecomment-754680777, as far as I can tell this PR has the same issues as https://github.com/lampepfl/dotty/pull/10996 and should be rejected for the same reason in favor of what I suggested in https://github.com/lampepfl/dotty/pull/10996#issuecomment-754719014

**Comment by changvvb** [comment]
I checked these cases and got the following results:
```scala
@deprecated("no CaseClass", "0.1") case class CaseClass(rgb: Int)

object Test {

  abstract class TestX

  val x = new TestX {
    val obj = new CaseClass(5) // warn
  }

  val y = Option(5).map(new CaseClass(_)) // warn
  val z = CaseClass(4)  // not warn but seems not related with this PR
}
```

But I agree that should not check synthetic flag only.

**Comment by changvvb** [comment]
@bishabosha @smarter Is it possible to check the `sourcePos` of ctx owners? we can check whether exists owner `sourcePos` that 
1.  is not `<no position>`
2. is not `sourcePos` of the deprecated symbol.

I think the user code must have a valid `sourcePos` right?

**Comment by changvvb** [comment]
> @bishabosha @smarter Is it possible to check the `sourcePos` of ctx owners? we can check whether exists owner `sourcePos` that
> 
> 1. is not `<no position>`
> 2. is not `sourcePos` of the deprecated symbol.
> 
> I think the user code must have a valid `sourcePos` right?

This is not a good solution because should disable not only `deprecated` warnings but also other warnings in synthetic scope.

**Pull request closed without merging**